src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs

```diff
@@ -0,0 +1,156 @@
+                // (6,12): error CS8107: Feature 'readonly structs' is not available in C# 7. Please use language version 7.1 or greater.
+                //     public readonly struct S2{}
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, "readonly").WithArguments("readonly structs", "7.1").WithLocation(6, 12),
+                // (8,5): error CS8107: Feature 'readonly structs' is not available in C# 7. Please use language version 7.1 or greater.
+                //     readonly public struct S3{}
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, "readonly").WithArguments("readonly structs", "7.1").WithLocation(8, 5)
+            );
+        }
+
+        [Fact]
+        public void ReadOnlyClassErr()
+        {
+            var text = @"
+class Program
+{
+    readonly class S1{}
+
+    public readonly delegate ref readonly int S2();
+
+    readonly public interface S3{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
+                // (4,20): error CS0106: The modifier 'readonly' is not valid for this item
+                //     readonly class S1{}
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "S1").WithArguments("readonly").WithLocation(4, 20),
+                // (6,47): error CS0106: The modifier 'readonly' is not valid for this item
+                //     public readonly delegate ref readonly int S2();
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "S2").WithArguments("readonly").WithLocation(6, 47),
+                // (8,31): error CS0106: The modifier 'readonly' is not valid for this item
+                //     readonly public interface S3{}
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "S3").WithArguments("readonly").WithLocation(8, 31)
+            );
+        }
+
+        [Fact]
+        public void ReadOnlyRefStruct()
+        {
+            var text = @"
+class Program
+{
+    readonly ref struct S1{}
+
+    unsafe readonly public ref struct S2{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll.WithAllowUnsafe(true));
+            comp.VerifyDiagnostics(
+            );
+        }
+
+        [Fact]
+        public void ReadOnlyStructPartial()
+        {
+            var text = @"
+class Program
+{
+    readonly partial struct S1{}
+
+    readonly partial struct S1{}
+
+    readonly partial ref struct S2{}
+
+    readonly partial ref struct S2{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
+            );
+        }
+
+        [Fact]
+        public void ReadOnlyStructPartialErr()
```

Comment:
Actually I did expect errors here, but we seem to not generally require modifiers to match. - You can state "unsafe" or "static" only on one part to have effect on all other parts. That could even be be intentional to support some source generation scenarios where generated parts would not need to restate modifiers. 

I will enter a bug to figure what to do with "ref" and "readonly", but will keep the behavior as is for now. And will change the name of the test as based on precedent I am no longer convinced if this should be an error.

---
In reply to: [118633262](https://github.com/dotnet/roslyn/pull/19713#discussion_r118633262) [](ancestors = 118633262)